---
title: "figure_3"
author: "Vincent Fischer"
date: "26 05 2023"
output:
  html_document:
    toc: true
    toc_float: true
    theme: spacelab
    code_folding: hide
    fig_width: 12
    df_print: paged
    number_sections: true
    

---

<style type="text/css">
.main-container {
  max-width: 75% !important;
  margin: auto;
}
</style>
```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(rtracklayer)
  library(GenomicRanges)
  library(ggplot2)
  library(ComplexHeatmap)
  library(chromVAR)
  library(motifmatchr)
  library(memes)
  library(MotifDb)
  library(universalmotif)
  library(TFBSTools)
  library(AnnotationHub)
  library(ggrepel)
  library(limma)
  library(SummarizedExperiment)
  library(cowplot)
})
theme_set(theme_bw())
BiocParallel::register(BPPARAM = BiocParallel::MulticoreParam(4,progressbar = TRUE))
set.seed(123)
```




# A
## load and prepare data
load motifs
```{r}
core <- read_meme(file = "../object_resources/HOCOMOCOv11_core_MOUSE_mono_meme_format.meme")
#full <- read_meme(file = "HOCOMOCOv11_full_MOUSE_mono_meme_format.meme")


# original function PL
summary_core <- universalmotif::summarise_motifs(core)
core <- setNames(object = core,nm = summary_core$name)
  
motifs <- core
modf <- data.frame(row.names=summary_core$name,
                     TF=gsub("_MOUSE.H11MO...+","",summary_core$name),
                     grade=gsub(".+\\.","",summary_core$name),
                     motif_rank = gsub("[^0123]+","",gsub(".*MO.",replacement = "",summary_core$name))) #gsub("[^ro]+", "", str1)
modf <- modf[order(modf$TF,modf$grade,modf$motif_rank),]
modf <- modf[!duplicated(modf$TF),]
modf <- modf[!modf$grade == "D",]
motifs <- setNames(universalmotif::convert_motifs(motifs[row.names(modf)]), modf$TF)
```

load reference genome fasta file (GRCm38)
```{r}
fa <- Rsamtools::FaFile("../object_resources/Mus_musculus.GRCm38.dna_sm.primary_assembly.fa")
```

Load NF peaks
```{r, fig.width=8.5, fig.height=4.8}
bl <- import("../object_resources/mm10.blacklist_chrM.bed")
seqlevelsStyle(bl) <- "ensembl"

nfp <- import("../object_resources/merged_NF_peaks.bed")
nfp <- nfp[!overlapsAny(nfp, bl)]
```
list bam files and get counts per peak
```{r, eval=TRUE}
bamfiles <- list.files("../Dex_Veh/aligned_full/", pattern="bam$", full=TRUE)
names(bamfiles) <- gsub("\\.bam$","",basename(bamfiles))
p <- resize(nfp[width(nfp)<250],width=250,fix="center")
sc <- getCounts(bamfiles, p, paired=TRUE)
saveRDS(sc, "../object_resources/figure_3_meta/p1.counts.rds")
sc <- addGCBias(sc, genome=fa)
```

```{r}
#sc <- readRDS("p1.counts.rds")
```


load motif archetypes and match to peaks
```{r, eval=TRUE}

moi <- readRDS("../object_resources/archetype_hits.GR.rds")
seqlevelsStyle(moi) <- "ensembl"
moi$type <- as.factor(moi$type)
o <- findOverlaps(rowRanges(sc), moi, minoverlap=5L)
o <- data.frame(peak=from(o), motif=as.integer(moi$type[to(o)]))
o <- o[!duplicated(o),]
library(Matrix)
m <- sparseMatrix(i=o[,1], j=o[,2])
colnames(m) <- levels(moi$type)
moix <- SummarizedExperiment(list(motifMatches=as(m, "lgCMatrix")), rowRanges=rowRanges(sc))
moix$name <- levels(moi$type)

```

compute deviation of motif activity
```{r, eval=TRUE}
set.seed(123)
dev <- computeDeviations(object=sc, annotations=moix,background_peaks = getBackgroundPeaks(sc,rowData(sc)$bias,niterations = 5000))
saveRDS(dev, "../object_resources/figure_3_meta/motifArchetypes_deviations_Dex_vs_Veh.rds")
#dev <- readRDS("../object_resources/figure_3_meta/motifArchetypes_deviations_Dex_vs_Veh.rds")
```

z-score distribution test
```{r, eval =FALSE}
d <- dplyr::bind_rows(lapply(as.list(as.data.frame(as.matrix(assays(dev)$z))),FUN=function(x) as.data.frame(as.matrix(x))), .id="sample")
ggplot(d, aes(V1, colour=sample)) + geom_density() + xlab("z-score")
```

normalization of z scores
```{r}
assays(dev)$norm <- scale(assays(dev)$z)
```

```{r, eval=FALSE}
d <- dplyr::bind_rows(lapply(as.list(as.data.frame(as.matrix(assays(dev)$norm))),FUN=function(x) as.data.frame(as.matrix(x))), .id="sample")
ggplot(d, aes(V1, colour=sample)) + geom_density() + xlab("norm")
```

## Generate Figure A
```{r, fig.height=3.5, fig.width=3}
dev$groups <- c("Dex","Dex","Dex","Veh","Veh")
dev$groups <-  relevel(factor(dev$groups), "Veh")
metadata(dev)$anno_colors <- list(groups=c(Veh="lightgrey", Dex="darkred"))
fit <- lmFit(assays(dev)$norm, model.matrix(~dev$groups))
head(mores <- topTable(eBayes(fit),number=Inf))


h <- sechm::sechm(dev[,as.integer(c(4,5,1,2,3))], row.names(mores)[mores$adj.P.Val<0.1], assayName = "norm", top_annotation= "groups", cluster_cols = FALSE, name = "norm. z")
arch <- readRDS("../object_resources/archetypes.rds")
arch <- arch[!duplicated(arch[,2]),]
row.names(arch) <- arch[,2]
mores$TFs <- arch[row.names(mores),"TFs"]
mores$label <- paste0(row.names(mores), " (", mores$TFs,")")

mores$label[1:2] <- c("NR/20 (ANDR/GCR/PRGR)", "HIF (ARNT/HIF1A/EPAS1)")

p <- ggplot(mores, aes(logFC, -log10(adj.P.Val), label=label)) + geom_point() + geom_text_repel(data=mores[mores$adj.P.Val<0.1,], hjust=0, nudge_x=0.2,force = 10) + xlab("activity norm. z-score logFC")+ggtitle("Motif activity at NFA regions")+ geom_hline(yintercept=-log10(0.05), linetype="dashed")+
geom_text(aes(2,-log10(0.05),label = "adjst.p-value = 0.05", vjust = -1.0),size = 3.5)+
  ylim(0,1.6)
diffMotif <- plot_grid(p, grid.grabExpr(draw(h)), nrow=1, rel_heights=c(3,2), scale=0.95,ncol = 2)
saveRDS(diffMotif,"../object_resources/figure_3_meta/p1.rds")
#diffMotif <- readRDS("p1_2.rds")

```

```{r,fig.width=11,fig.height=4}
diffMotif
```




# B

## loading data

```{r}
bamfiles_Dex <- list.files(path = "../Dex_Veh/aligned_full/",pattern = "^D(.*).bam$",full.names = TRUE)
bamfiles_Veh <- list.files(path = "../Dex_Veh/aligned_full/",pattern = "^V(.*).bam$",full.names = TRUE)
names(bamfiles_Dex) <- c("Dex1","Dex2","Dex3")
names(bamfiles_Veh) <- c("Veh1","Veh2")
```

read in 5fC sites
```{r}
sites_5fC <- readRDS("../object_resources/5fC/sites_5fC.rds")
```

sort peaks applying filter peaks later; necessary to get only peaks with at least on fragment
```{r}
sort(sites_5fC)
sites_5fC
```

read in counts of Dex vs Veh at 5fC sites
```{r}
sc <- getCounts(c(bamfiles_Veh,bamfiles_Dex), reduce(resize(sites_5fC,width = 200,fix = "center"),), paired=TRUE)
saveRDS(sc,"../object_resources/figure_3_meta/sc_DexVeh_5fC.rds")
```

add GC bias
```{r}
#sc <- readRDS("../object_resources/figure_3_meta/sc_DexVeh_5fC.rds")
sc <- addGCBias(sc, genome=fa)
```


```{r}
# filter needed for 'computeDeviations'
counts_filtered <- filterPeaks(sc, non_overlapping = FALSE)

#motifs <- do.call(what = "PFMatrixList", lapply(getNonRedundantMotifs(), class="TFBSTools-PFMatrix", FUN= convert_motifs))
motifs2 <- do.call(PFMatrixList, lapply(motifs, class="TFBSTools-PFMatrix", FUN=convert_motifs))
motif_ix <- matchMotifs(motifs2, counts_filtered, genome=fa)
```


```{r}
set.seed(123)
dev <- computeDeviations(object=counts_filtered, annotations=motif_ix,background_peaks = getBackgroundPeaks(counts_filtered,rowData(counts_filtered)$bias,niterations = 5000))
```

```{r}
dev$groups <- c("Veh","Veh","Dex","Dex","Dex")
dev$groups <-  relevel(factor(dev$groups), "Veh")
saveRDS(dev,"../object_resources/figure_3_meta/5fC_dev.rds")
dev <- readRDS("../object_resources/figure_3_meta/5fC_dev.rds")
```

z-score distribution test
```{r,eval=FALSE}
d <- dplyr::bind_rows(lapply(as.list(as.data.frame(as.matrix(assays(dev)$z))),FUN=function(x) as.data.frame(as.matrix(x))), .id="sample")
ggplot(d, aes(V1, colour=sample)) + geom_density() + xlab("z-score")
```
normalization of z scores
```{r}
assays(dev)$norm <- scale(assays(dev)$z)
```

```{r, eval=FALSE}
d <- dplyr::bind_rows(lapply(as.list(as.data.frame(as.matrix(assays(dev)$norm))),FUN=function(x) as.data.frame(as.matrix(x))), .id="sample")
ggplot(d, aes(V1, colour=sample)) + geom_density() + xlab("z-score")
```

lmFit
```{r}
fit <- lmFit(assays(dev)$norm, model.matrix(~dev$groups))
head(mores <- topTable(eBayes(fit),number=Inf),395)
metadata(dev)$anno_colors <- list(groups=c(Veh="darkgrey", Dex="darkred"))
```

## Generate figure B
```{r}
h <- sechm::sechm(dev, row.names(mores)[1:2], assayName = "norm", top_annotation="groups", breaks=0.99, name = "norm. z")

```

```{r,fig.width=10}
w <- c(1:2)
w
gg <-ggplot(mores, aes(logFC, -log10(adj.P.Val ))) +
  geom_point() +
  geom_text_repel(data=cbind(gene=row.names(mores)[w], mores[w,]), aes(label=gene),max.overlaps = 20,nudge_y= -0.05)+
  #geom_hline(yintercept=-log10(0.05), linetype="dashed")+
  #geom_text(aes(-2,-log10(0.05),label = "adjst.p-value = 0.05", vjust = -1.5,),size = 3.5)+
  ggtitle("Motif activity at 5fC sites")+
  xlab("activity norm. z-score logFC")
gg

```

```{r}
diffMotif_5fC <- plot_grid(gg, grid.grabExpr(draw(h)), nrow=1, rel_heights=c(3,2), scale=0.95,ncol = 2)
saveRDS(diffMotif_5fC,"p2.rds")
#diffMotif_5fC <- readRDS("p2.rds")
```

```{r,fig.width=11,fig.height=4}
diffMotif_5fC
```

# C

load and prepare data
```{r}
bw_ATAC_mono <- list.files(path = "../Dex_Veh/tracks/", pattern ="*Veh.(.*)mono.bw",full.names = TRUE )
bw_ATAC_nf <- list.files(path = "../Dex_Veh/tracks/", pattern ="*Veh.(.*)NF.bw",full.names = TRUE )
bw_ATAC_full <- list.files(path = "../Dex_Veh/tracks/", pattern ="*Veh.(.*)full.bw",full.names = TRUE )

names(bw_ATAC_mono) <- c("ATAC_Mono_1","ATAC_Mono_2")
names(bw_ATAC_nf) <- c("ATAC_NF_1","ATAC_NF_2")
names(bw_ATAC_full) <- c("ATAC_Full_1","ATAC_Full_2")
```

load TSS with 5fC in proximity
```{r}
TSS_2KB_5fC<-readRDS("../object_resources/5fC/TSS_2KB_5fC.rds")
```
 
generation of normalization factors
```{r}
bg_nf <- bwNormFactors(c(bw_ATAC_nf,bw_ATAC_mono,bw_ATAC_full), method = "background")
```

generate matrix
```{r}
m_5fC <- signal2Matrix(c(bw_ATAC_nf,bw_ATAC_mono,bw_ATAC_full),w = 10L, regions = TSS_2KB_5fC)
m_5fC_bg <- rescaleSignalMatrices(m_5fC,bg_nf)
```

cluster by kmeans to get cluster with and without accessibility
```{r}
set.seed(123)  # to ensure that it gives the same results everytime
cl <- clusterSignalMatrices(m_5fC_bg, k=2)
table(cl)
cl
```

subsetting TSS_2KB_5fC for cluster 1
```{r}
cl_TSS_2KB_5fC <- TSS_2KB_5fC[cl == 1 ,]
cl_TSS_2KB_5fC
```

load bam files
```{r}
bamfiles_Dex <- list.files(path = "../Dex_Veh/aligned_full/",pattern = "^D(.*).bam$",full.names = TRUE)
bamfiles_Veh <- list.files(path = "../Dex_Veh/aligned_full/",pattern = "^V(.*).bam$",full.names = TRUE)
names(bamfiles_Dex) <- c("Dex1","Dex2","Dex3")
names(bamfiles_Veh) <- c("Veh1","Veh2")
```

count reads, match motifs and compute deviations
```{r}
sc <- chromVAR::getCounts(c(bamfiles_Veh,bamfiles_Dex), peaks = reduce(resize(cl_TSS_2KB_5fC,width = 200,fix = "center")), paired=TRUE)

sc <- addGCBias(sc, genome=fa)
saveRDS(sc,"../object_resources/figure_3_meta/sc_TSS_5fC_cl.rds")


#sc <- readRDS("../object_resources/figure_3_meta/sc_TSS_5fC_cl.rds")

# filter needed for 'computeDeviations'
counts_filtered <- filterPeaks(sc, non_overlapping = FALSE)
```

read in archetype motifs
```{r}
#moi <- readRDS("../object_resources//archetype_hits.GR.rds")
seqlevelsStyle(moi) <- "ensembl"
moi$type <- as.factor(moi$type)
o <- findOverlaps(rowRanges(counts_filtered), moi, minoverlap=5L)
o <- data.frame(peak=from(o), motif=as.integer(moi$type[to(o)]))
o <- o[!duplicated(o),]
library(Matrix)
m <- sparseMatrix(i=o[,1], j=o[,2])
colnames(m) <- levels(moi$type)
moix <- SummarizedExperiment(list(motifMatches=as(m, "lgCMatrix")), rowRanges=rowRanges(counts_filtered))
moix$name <- levels(moi$type)
```


```{r}
set.seed(123)
dev_arch <- computeDeviations(object=counts_filtered, annotations=moix,background_peaks = getBackgroundPeaks(counts_filtered,rowData(counts_filtered)$bias,niterations = 5000))
saveRDS(dev_arch, "../object_resources/figure_3_meta/dev_with_arch_TSS_clustered2.rds")
#dev_arch <- readRDS("../object_resources/figure_3_meta/dev_with_arch_TSS_clustered2.rds")
```

```{r}
dev_arch$groups <- c("Veh","Veh","Dex","Dex","Dex")
dev_arch$groups <-  relevel(factor(dev_arch$groups), "Veh")
```

z-score distribution test
```{r, eval =FALSE}
d <- dplyr::bind_rows(lapply(as.list(as.data.frame(as.matrix(assays(dev_arch)$z))),FUN=function(x) as.data.frame(as.matrix(x))), .id="sample")
ggplot(d, aes(V1, colour=sample)) + geom_density() + xlab("z-score")
```

normalization of z scores
```{r,eval=TRUE}
assays(dev_arch)$norm <- scale(assays(dev_arch)$z)
```

```{r,eval=FALSE}
d <- dplyr::bind_rows(lapply(as.list(as.data.frame(as.matrix(assays(dev_arch)$norm))),FUN=function(x) as.data.frame(as.matrix(x))), .id="sample")
ggplot(d, aes(V1, colour=sample)) + geom_density() + xlab("norm. z-score")
```


```{r}
metadata(dev_arch)$anno_colors <- list(groups=c(Veh="darkgrey", Dex="darkred"))
fit <- lmFit(assays(dev_arch)$norm, model.matrix(~dev_arch$groups))#
head(mores <- topTable(eBayes(fit),number=Inf))

arch <- readRDS("../object_resources/archetypes.rds")
arch <- arch[!duplicated(arch[,2]),]
row.names(arch) <- arch[,2]
mores$TFs <- arch[row.names(mores),"TFs"]
mores$label <- paste0(row.names(mores), " (", mores$TFs,")")
```

## generate figure C
```{r}
h <- sechm::sechm(dev_arch, row.names(mores)[1:3], assayName = "norm", top_annotation="groups", breaks=0.99,name = "norm. z")
```

```{r}
gg <- ggplot(mores, aes(logFC, -log10(adj.P.Val
))) + 
  geom_point() + geom_text_repel(data=head(cbind(gene=row.names(mores), mores),1), aes(label=gene),max.overlaps = 20)+
  labs(x="activity norm. z-score logFC")+ xlim(c(-5,3))+
  #theme(text = element_text(size = 12))+
  ggtitle("Motif activity at accessible TSS with 5fC")
gg
```

```{r}
diffMotif_cl_TSS_5fC_2cluster <- plot_grid(gg, grid.grabExpr(draw(h)), nrow=1, rel_heights=c(3,2), scale=0.95,ncol = 2)
saveRDS(diffMotif_cl_TSS_5fC_2cluster,"../object_resources/figure_3_meta/p3_2cluster.rds")
#diffMotif_cl_TSS_5fC_2cluster <- readRDS("../object_resources/figure_3_meta/p3_2cluster.rds")
```


# Joint
```{r, fig.width=9, fig.height=8,eval=TRUE}
#plot_grid(grid.grabExpr(ComplexHeatmap::draw(p1)), nrow=1, rel_widths=c(5,3))
pp <- plot_grid(diffMotif,diffMotif_5fC,diffMotif_cl_TSS_5fC_2cluster, nrow = 3, labels = c("A","B","C"))
pp

# now add the title
title <- ggdraw() + 
  draw_label(
    "Figure 3",
    fontface = 'bold',
    x = 0,
    hjust = 0
  ) +
  theme(
    # add margin on the left of the drawing canvas,
    # so title is aligned with left edge of first plot
    plot.margin = margin(0, 0, 0, 7)
  )

pp2 <- plot_grid(
  title, pp,
  ncol = 1,
  # rel_heights values control vertical title margins
  rel_heights = c(0.1, 2)
)

pp2

pdf("figure_3_label.pdf", width=9, height=8)
pp2
dev.off()
```

```{r}
sessionInfo()
```

